### PR TITLE
Add support for v1.PodLogOptions

### PR DESCRIPTION
**Pull request by ldx**
Check if "follow" is requested for container logs, and pass it to the
provider so tailing logs can be supported.

### CODE REVIEW

**Comment by cpuguy83** [comment]
From what I can tell in the kubelet, there is no supported query param for `follow`.
https://github.com/kubernetes/kubernetes/blob/15c3f1b119ff4b4073c86df05c819a9672b80d68/pkg/kubelet/server/server.go#L542-L583

**Comment by ldx** [comment]
There is: https://github.com/kubernetes/kubernetes/blob/15c3f1b119ff4b4073c86df05c819a9672b80d68/staging/src/k8s.io/api/core/v1/types.go#L4808

If you use `kubectl logs -f`, `follow=true` will be present in the API request query.

**Comment by cpuguy83** [comment]
Right so we need to decode the log options from the request.

**Comment by ldx** [comment]
> Right so we need to decode the log options from the request.

Yes, the same way as with `tailLines`.

Thinking a bit more about this: we could even add support for all fields in `v1.PodLogOptions` here in one go. Let me know, and I can refactor this PR to do that.

File: node/api/logs.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ type ContainerLogOpts struct {
 	Since      time.Duration
 	LimitBytes int
 	Timestamps bool
+	Follow     bool
```
Comment (cwdsuzhou) [inline node/api/logs.go]: we  can not just add Follow option, since the current implementation actually do not support it

Comment (ldx) [reply node/api/logs.go]: @cwdsuzhou you mean providers don't support it? Ours definitely can once this field is in the struct.

**Comment by cpuguy83** [comment]
Yes let's support v1.PodLogOptions and convert that to our log config.

Thank you!

**Comment by ldx** [comment]
@cpuguy83 I refactored this to include all fields from v1.PodLogOptions.

File: node/api/logs.go (hunk 1 above)
Comment (cwdsuzhou) [reply node/api/logs.go]: Maybe something wrong with mysql implementation. But when I change some logic of `vk`, it works.

**Comment by ldx** [comment]
@cpuguy83 ping - anything else I can do?

File: node/api/logs.go [hunk 2]
```diff
@@ -33,10 +35,71 @@ type ContainerLogsHandlerFunc func(ctx context.Context, namespace, podName, cont
...
+}
+
+func parseLogOptions(q url.Values) (opts ContainerLogOpts, err error) {
+	if tailLines := q.Get("tailLines"); tailLines != "" {
+		opts.Tail, err = strconv.Atoi(tailLines)
+		if err != nil {
+			return opts, errdefs.AsInvalidInput(errors.Wrap(err, "could not parse \"tailLines\""))
+		}
+		if opts.Tail < 0 {
+			return opts, errdefs.InvalidInput(fmt.Sprintf("\"tailLines\" is %d", opts.Tail))
```
Comment (cpuguy83) [inline node/api/logs.go:54]: Can you use `InvalidInputf` instead of `fmt.Sprintf`?

Comment (cpuguy83) [reply node/api/logs.go:54]: There are a few of these.

Comment (ldx) [reply node/api/logs.go:54]: Sure, let me fix this :+1:

Comment (ldx) [reply node/api/logs.go]: @cpuguy83 done.

**Pull request merged**